src/Analyzers/CSharp/Analyzers/ConvertProgram/ConvertProgramAnalysis_TopLevelStatements.cs

```diff
@@ -0,0 +1,159 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Analyzers.ConvertProgram
+{
+    internal static partial class ConvertProgramAnalysis
+    {
+        public static bool CanOfferUseTopLevelStatements(CodeStyleOption2<bool> option, bool forAnalyzer)
+        {
+            var userPrefersTopLevelStatements = option.Value == true;
+            var analyzerDisabled = option.Notification.Severity == ReportDiagnostic.Suppress;
+            var forRefactoring = !forAnalyzer;
+
+            // If the user likes top level statements, then we offer to convert to them from the diagnostic analyzer.
+            // If the user prefers Program.Main then we offer to use top-level-statements from the refactoring provider.
+            // If the analyzer is disabled completely, the refactoring is enabled in both directions.
+            var canOffer = userPrefersTopLevelStatements == forAnalyzer || (forRefactoring && analyzerDisabled);
+            return canOffer;
+        }
+
+        public static Location GetUseTopLevelStatementsDiagnosticLocation(MethodDeclarationSyntax methodDeclaration, bool isHidden)
+        {
+            // if the diagnostic is hidden, show it anywhere on the main method. Otherwise, just put the diagnostic on
+            // the the 'Main' identifier.
+            return isHidden ? methodDeclaration.GetLocation() : methodDeclaration.Identifier.GetLocation();
+        }
+
+        public static string? GetMainTypeName(Compilation compilation)
+        {
+            var mainTypeFullName = compilation.Options.MainTypeName;
+            var mainTypeName = mainTypeFullName?.Split('.').Last();
+            return mainTypeName;
+        }
+
+        public static bool IsProgramMainMethod(
+            SemanticModel semanticModel,
+            MethodDeclarationSyntax methodDeclaration,
+            string? mainTypeName,
+            CancellationToken cancellationToken,
+            out bool canConvertToTopLevelStatements)
+        {
+            canConvertToTopLevelStatements = false;
+
+            // Quick syntactic checks to allow us to avoid most methods.  We basically filter out anything that isn't
+            // `static Main` immediately.
+            if (!methodDeclaration.Modifiers.Any(SyntaxKind.StaticKeyword) ||
+                methodDeclaration.TypeParameterList is not null ||
+                methodDeclaration.Identifier.ValueText != WellKnownMemberNames.EntryPointMethodName ||
+                methodDeclaration.Parent is not TypeDeclarationSyntax containingTypeDeclaration ||
+                methodDeclaration.Body == null)
```

Comment:
Expression body?

Reply:
Correct. Not supported. I'll add comment. It's just much simpler this way.